### ISSUE TITLE
Upgrade to pac4j v1.8

### ISSUE THREAD

**Issue by leleuj** [jooby-project/jooby#214]
We just released pac4j v1.8 and as Jooby uses it for authentication, it should be upgraded.

pac4j v1.8 is a major version as it is now a full security engine and not only an authentication one for Facebook, Twitter...: it now supports REST/web services as well as authorizations. Which allows implementations (like jooby-pac4j) to become full security libraries.
Configuration can now be handled at the pac4j level (-> https://github.com/pac4j/pac4j/tree/master/pac4j-core/src/main/java/org/pac4j/core/config) as well as profile storage (-> https://github.com/pac4j/pac4j/blob/master/pac4j-core/src/main/java/org/pac4j/core/profile/ProfileManager.java).

The idea with pac4j v1.8 is not only to have implementations which are close, but to provide guidelines on how to implement pac4j and have the **same algorithm** implemented for all frameworks/tools, making evolutions and adoption easier.

If you read: https://github.com/pac4j/pac4j,

There are now two kinds of clients (= authentication mechanims):
- **indirect / stateful clients** are for UI when the user authenticates once at an external provider (like Facebook, a CAS server...) or via a local form (or basic auth popup)
- **direct / stateless clients** are for web services when credentials (like basic auth, tokens...) are passed for each HTTP request.

To define your security configuration, gather all your authentication mechanisms = [**clients**](https://github.com/pac4j/pac4j/wiki/Clients) via the `Clients` class (to share the same callback url). Also define your [**authorizers**](https://github.com/pac4j/pac4j/wiki/Authorizers) to check authorizations and aggregate both (clients and authorizers) on the `Config`.
Notice you may also use the `ConfigSingleton` object to keep one instance of your configuration and share it among the different components (if you don't have any dependency injection capability). You can also use the `ConfigFactory` to build you configuration if no other mean is available.

To secure your Java web application, **the reference implementation is to create two "filters"**: **one to protect urls**, **the other one to receive callbacks** for stateful authentication processes (indirect clients):
1) **For your protection "filter", it must be based on two String parameters: `clientName`** (list of clients used for authentication) **and `authorizerName`** (list of authorizers to check authorizations) and **use the following logic: loop on direct clients for authentication then check the user profile and authorizations**
2) **For your callback "filter", get the credentials and the user profile on the callback url**.

We moved things really far as you can see that both implementations between J2E and Play for example are very very close: https://github.com/pac4j/j2e-pac4j/blob/master/src/main/java/org/pac4j/j2e/filter/RequiresAuthenticationFilter.java#L91 vs https://github.com/pac4j/play-pac4j/blob/master/play-pac4j-java/src/main/java/org/pac4j/play/java/RequiresAuthenticationAction.java#L95.

Even the README are organized the same way with the same texts for common parts. The associated demos have been upgraded and that should be done for: https://github.com/pac4j/jooby-pac4j-demo as well. The README of the demo can be copy/pasted from the one for J2E and updated.
There is even a specific logo for the jooby-pac4j implementation: https://pac4j.github.io/pac4j/img/logo-jooby.png which should appear at the top of the README demo and of the implementation documentation: http://jooby.org/doc/pac4j/

If you have any question or need any help, don't hesitate.

**Comment by jknack** [comment]
Hello @leleuj,

I'm about to update jooby-pac4j integration and I do have a question about: https://github.com/pac4j/pac4j/wiki/Authorizers

Let's say I have an app with a ```form``` client, this app always requires authentication, like:

```java
config.addAuthorizer("user", new RequireAnyRoleAuthorizer("user"));
```

But also, there is an admin section at ```/admin/**```, this url requires an ```admin``` role. My question is: How do I add or specify this via authorizer API?

Thanks

**Comment by leleuj** [comment]
Great!

If you only require your user to be authenticated, you don't need to define any Authorizer, the protection filter will only have a clientName parameter to reference the client used for authentication (here: the FormClient). You could even specify a blank or no clientName parameter, it should mean that the url is protected but no authentication process will be played (it must occured on another url).

If your user must have an admin role on some url, you'll need to define a RequireAnyRoleAuthorizer with this role in the Config and reference it with the authorizerName parameter in your protection filter (in addition to the clientName parameter referencing the client used for authentication).

The security configuration (clients for authentication and authorizers for authorizations) is centralized and you may use it in the constructor of your protection filter or inject it if dependency injection is supported by Jooby or shared it via the ConfigSingleton class.

**Comment by jknack** [comment]
got it!

have everything more or less working, need to work on tests.

everything seems to be working fine, but I do have a suggestion/question: 

Think ```ProfileManager.get(boolean useSession)``` is kind of limited. In current version I do have: ```AuthStore.get(String profileId)``` which I think is more flexible and isn't limited to just save the profile in the Session (usually memory).

In Jooby integration, there is one ```AuthStore``` implementation which effectively and by default use the HTTP Session to save the data, but users can override/set any other kind of storage.

I'm going to update everything in jooby, but keep ```AuthStore``` and I avoid ```ProfileManager``` for now.

Thoughts?

**Comment by leleuj** [comment]
Your suggestion makes sense. Though, I think that this topic must be addressed in a more global way: if you don't want to use the HTTP session for the user profile, it's certainly a more general choice for your app. So other elements (like tokens, states...), just to talk about pac4j, will certainly go to the new store as well.

To have another session implementation, things must be changed at the context level (which deals with the request, response and session). For Play framework (where the pac4j session used is a cache and not the default Play session which goes into a cookie), there is a concept of `DataStore` and a  default `CacheStore` implementation at the `PlayContext` level: https://github.com/pac4j/play-pac4j/blob/master/play-pac4j-java/src/main/java/org/pac4j/play/PlayWebContext.java#L50. It's certainly too specific, but it's a first draft.

Thus, the `useSession` targets the current session implementation and it's necessary to distinguish between a session (stateful) profile and a request (stateless) profile.

Does it make sense?

Apart from this store change capability, the contracts in the `AuthStore` and in the `ProfileManager` are very close, except the `get` method. Why do you need the identifier to get the profile?

**Comment by jknack** [comment]
They are close, indeed. Feel direct/stateless auth plays better/nice with an ID. So, clients give us an ID (or similar) and we go directly the the profile manager and get it.

**Comment by leleuj** [comment]
I'm not sure to follow you on the "clients give us an ID". As a matter of facts, clients provide profiles with identifiers and attributes.
In all cases, these are just guidelines which are better to follow to help pac4j users go from one technology to another.

One good thing to do would be have a full demo for Jooby: https://github.com/pac4j/jooby-pac4j-demo like the reference demo for J2E: https://github.com/pac4j/j2e-pac4j-demo

**Comment by leleuj** [comment]
Do you plan to upgrade the demo as well?

**Comment by jknack** [comment]
v0.12.0 was released it last night. demo was upgraded following the jee demo, with v0.12.0 out, will update demo and push it later today.

jooby has a new website: http://jooby.org and logo ;)

**Comment by leleuj** [comment]
Great! Your new website and logo really rock now. I'll upgrade the pac4j-jooby logo soon.
Let me know if you need help to update the demo. I'll be soon back to you as I released pac4j v1.8.1 (very easy upgrade).

**Comment by jknack** [comment]
sent you a pull request: https://github.com/pac4j/jooby-pac4j-demo/pull/3

demo has the same clients as ```j2ee``` but UI is slightly different, still the main concept is exactly the same for both.

**Comment by leleuj** [comment]
Thanks again for the updated demo. I just updated things myself to go through all use cases: https://github.com/pac4j/jooby-pac4j-demo/commit/8c7a74e8e8fbf3caa1d424cdfc04924994009e23

1) I always create a specific logo for pac4j specific implementation: https://pac4j.github.io/pac4j/img/logo-jooby.png Doesn't that bother you?  Do you have an official jooby logo which I can use?

2) I haven't been able to find in the documenation how to get the config here: https://github.com/pac4j/jooby-pac4j-demo/blob/master/src/main/java/org/pac4j/demo/jooby/App.java#L46

3) I have added a use case for JWT token authentication, but I always get a 401. My flow: login with form, generate a token (specific link), logout, call the /rest-jwt url with token=*generated_token*

4) How can I define multiple clients (web services authentication use case) on the same url (like basic auth + JWT token)?

Thanks.

**Comment by jknack** [comment]
Hi,

1. no problem, you can use that one or this one: https://github.com/jooby-project/jooby/blob/gh-pages/resources/images/apps/android-chrome.png. All the logos are here: https://github.com/jooby-project/jooby/tree/gh-pages/resources/images

2. any service is accessible by calling ```req.require```.

```
Config conf = req.require(Config.class);
```

3. will review and try it later, it should work

4. This is a good one, think isn't possible to do this today. Can you raise a bug for this?

Thank you, @leleuj

**Comment by leleuj** [comment]
1. I finally used the named logo to be consistent with other logos, it looks correct now: https://pac4j.github.io/pac4j/img/logo-jooby.png
2. I sucessfully updated the demo
3. Waiting for your feedback
4. I just opened: https://github.com/jooby-project/jooby/issues/241

I'm used to see the general syntax of pac4j which works everywhere (a global `Config` with clients and authorizers, some "requires authentication" handler...), but the truth is that the Jooby syntax is terrific (it really competes with Ratpack or Sparkjava).

**Comment by jknack** [comment]
3. yea wait, but to be honest will probably get back to you next week

Thank you! API for pac4j is clean and rich!!!
I put a lot of energy designing API in general. By design, Jooby won't add boundaries or limits to a library, key concept is to simplify bootstrap and configuration as much as possible.

I tried to put these thoughts in the doc too: http://jooby.org/doc/#modules

**Comment by jknack** [comment]
here is the fix for 3: #245

**Comment by leleuj** [comment]
Great! Thanks
